### ISSUE TITLE
ConvertToHeifImage() using a byte array

### ISSUE THREAD

**Issue by pol2095** [0xC0000054/libheif-sharp-samples#1]
Hello,

I have 2 questions 
when I use 
`public HeifImage ConvertToHeifImage(Image<Rgb24> image)`
how to convert a jpg byte array to `Image<Rgb24>`

I tried
```
private byte[] byteArrayToImage(byte[] byteArrayIn)
{
	using (MemoryStream ms = new MemoryStream(byteArrayIn))
	{
		System.Drawing.Image img = System.Drawing.Image.FromStream(ms);
		SixLabors.ImageSharp.Image image = SixLabors.ImageSharp.Image.Load(ms);
		return ConvertToHeifImage(image);
	}
}
```
but how convert SixLabors.ImageSharp.Image to `SixLabors.ImageSharp.Image<Rgb24>` ?

and I need to convert LibHeifSharp.HeifImage to a byte array ?

Thanks

**Comment by 0xC0000054** [comment]
> but how convert SixLabors.ImageSharp.Image to SixLabors.ImageSharp.Image<Rgb24> ?

You need to use the `SixLabors.ImageSharp.Image.Load<Rgb24>` generic method.

> and I need to convert LibHeifSharp.HeifImage to a byte array ?

You can use the `WriteToStream` method instead of `WriteToFile`.

```C#

private byte[] SaveHeifImageToByteArray(HeifContext image)
{
	byte[] bytes;

    using (MemoryStream stream = new MemoryStream())
    {
        image.WriteToStream(stream);
        bytes = stream.ToArray();
    }

    return bytes;
}
```

**Comment by pol2095** [comment]
Thanks but the sample return a HeifImage, how to convert to HeifContext  ?

**Comment by pol2095** [comment]
Thanks, but I obtain a HeifImage, how to convert to HeifContext ?

**Comment by 0xC0000054** [comment]
> Thanks, but I obtain a HeifImage, how to convert to HeifContext ?

The `HeifImage` is the decoded image, you have to create a `HeifContext` and encode the image.
Something like:

```c#
private byte[] byteArrayToImage(byte[] byteArrayIn, int quality)
{
	SixLabors.ImageSharp.Image<Rgb24> inputImage = SixLabors.ImageSharp.Image.Load<Rgb24>(byteArrayIn);
	SixLabors.ImageSharp.Metadata.ImageMetadata metadata = inputImage.Metadata;

	HeifImage heifImage = ConvertToHeifImage(inputImage);

	// Set the HeifImage color profile information.
	if (metadata.IccProfile != null)
	{
		heifImage.ColorProfile = new HeifIccColorProfile(metadata.IccProfile.ToByteArray());
	}
	else
	{
		heifImage.ColorProfile = new HeifNclxColorProfile(ColorPrimaries.BT709,
													 TransferCharacteristics.Srgb,
													 MatrixCoefficients.BT601,
													 fullRange: true);
	}

	byte[] byteArrayOut = null;

	using (HeifContext context = new HeifContext())
	{
		HeifEncoder encoder = context.GetEncoder(HeifCompressionFormat.Hevc);

		encoder.SetLossyQuality(quality);

		using (HeifImageHandle imageHandle = context.EncodeImageAndReturnHandle(heifImage, encoder))
		{
			if (metadata.ExifProfile != null)
			{
				context.AddExifMetadata(imageHandle, metadata.ExifProfile.ToByteArray());
			}
		}

		using (System.IO.MemoryStream stream = new System.IO.MemoryStream())
		{
			context.WriteToStream(stream);
			byteArrayOut = stream.ToArray();
		}
	}

	return byteArrayOut;
}
```

**Comment by pol2095** [comment]
Thanks, but I have a problem

```
if (!LibHeifInfo.HaveEncoder(HeifCompressionFormat.Hevc))
{
    Console.WriteLine("No HEVC encoder available.");
}
```

why I have No HEVC encoder available ?

**Comment by 0xC0000054** [comment]
> why I have No HEVC encoder available ?

The libheif binary you are using was not compiled with HEVC support.

**Comment by pol2095** [comment]
I use [strukturag/libheif](https://github.com/strukturag/libheif), I obtain a dll using VS 2019 size of "x64-Release" 832 ko ?

**Comment by pol2095** [comment]
How to compile [libheif](https://github.com/strukturag/libheif) with HEVC support ?

**Comment by 0xC0000054** [comment]
> How to compile [libheif](https://github.com/strukturag/libheif) with HEVC support ?

As stated in the libheif readme, the easiest method to compile libheif on Windows is to use [Vcpkg](https://github.com/microsoft/vcpkg).

**Comment by pol2095** [comment]
The HEVC support is-it in [libde265](https://github.com/strukturag/libde265) ?

**Comment by 0xC0000054** [comment]
> The HEVC support is-it in [libde265](https://github.com/strukturag/libde265) ?

That is the HEVC decoder, the encoder is x265.

**Comment by pol2095** [comment]
I added the dll x265 but not work.

**Comment by pol2095** [comment]
My app work using a dll x64 but hen I create a dll x86, I have this message error :
`System.AccessViolationException: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.'`

**Comment by pol2095** [comment]
My app work now using a libheif.dll x64 but when I create a libheif.dll x86, I have this message error :

```
System.AccessViolationException: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   … LibHeifSharp.Interop.LibHeifNative.heif_context_write(SafeHeifContext context, SafeHandle writer, IntPtr userData)
   … LibHeifSharp.HeifContext.WriteToStreamInternal(Stream stream)
   … LibHeifSharp.HeifContext.WriteToStream(Stream stream)
```

**Comment by 0xC0000054** [comment]
I recently fixed a calling convention issue with the write callback, but it is still passing the parameters in the wrong order on X86.

**Comment by pol2095** [comment]
Good, is-it possible to fix it ?

**Comment by 0xC0000054** [comment]
> is-it possible to fix it ?

I do not know.
I tried to reproduce that bug in a test app, and I was not able to.

The issue is that the `write` callback in the `heif_writer` structure looses the first parameter when it is marshaled to managed code on X86.

The `heif_writer` [write callback method](https://github.com/strukturag/libheif/blob/667eeabb553ce73094eb29faea3f31fb8610fec2/libheif/heif.h#L1015) looks like this:

```c++
struct heif_error (* write)(struct heif_context* ctx,
                              const void* data,
                              size_t size,
                              void* userdata);
```

My managed code represents that callback using the following [delegate](https://github.com/0xC0000054/libheif-sharp/blob/de4545eb7643c3e3f53ee123c96defb74316b961/src/Interop/LibHeif/IO/heif_writer.cs#L28):

```c#
internal delegate heif_error WriteDelegate(IntPtr ctx, IntPtr data, UIntPtr size, IntPtr userData);
```

Both the native callback and my managed delegate are using the CDecl calling convention.
This works correctly on X64, but for some reason ox X86 the reverse P/Invoke call marshaling is incorrect.

The `data` parameter is marshaled as the `ctx` parameter, the `size` parameter is marshaled as the `data` parameter and the `userData` parameter is marshaled as the `size` parameter.
The only reason it does not crash is that the `userData` parameter is 0.

**Comment by pol2095** [comment]
you can't reproduce this bug using a x86 libheif.dll ?

**Comment by 0xC0000054** [comment]
> you can't reproduce this bug using a x86 libheif.dll ?

I can reproduce it with an x86 libheif dll, but not with a new C++ dll that uses similar code to what libheif is using.

**Comment by pol2095** [comment]
Is-it possible to ask [strukturag/libheif](https://github.com/strukturag/libheif) about this ?

**Comment by 0xC0000054** [comment]
> Is-it possible to ask strukturag/libheif about this?

It is not a bug in libheif, it is a bug in the .NET Runtime.

The bug occurs on the X86 version of .NET Core 3.1.10 (and possibly other .NET Core versions).
It does not occur on the X86 versions of .NET 5.0 or the .NET Framework version 4.8.

If I can get a simple demo working, I will report this issue to the [.NET Runtime](https://github.com/dotnet/runtime) repository.

**Comment by pol2095** [comment]
I use a project .NET Framework, the update 2.0.2 work now for x86, Thanks.

**Comment by 0xC0000054** [comment]
> I use a project .NET Framework, the update 2.0.2 work now for x86, Thanks.

That is good to hear.
Closing this issue.

**Comment by pol2095** [comment]
I created the dlls using vcpkg and it doesn't work for x86 and x64, I have this error

```
System.AccessViolationException: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   … LibHeifSharp.Interop.LibHeifNative.heif_context_encode_image(SafeHeifContext context, SafeHeifImage image, SafeHeifEncoder encoder, IntPtr options_MustBeZero, SafeHeifImageHandle& imageHandle)
   … LibHeifSharp.HeifContext.EncodeImageAndReturnHandle(HeifImage image, HeifEncoder encoder, HeifEncodingOptions options)
```

when I create the dll with CMake it work ?

**Comment by 0xC0000054** [comment]
> when I create the dll with CMake it work ?

That access violation is occurring when libheif attempts to get the version string that the x265 DLL exports.
It works when x265 is statically linked to libheif.

It probably has something to do with the fact that x265 exports the string as a pointer, instead of using a get method.
Looks like I have another bug report to file. 😄

**Comment by 0xC0000054** [comment]
Updating the Vcpkg version of libheif to 1.10.0 fixed that crash.
I submitted a [PR](https://github.com/microsoft/vcpkg/pull/15416) to the Vcpkg repo, hopefully it will be merged soon.

You could also copy the changes into your local Vcpkg port of libheif.

**Comment by KWEngine** [comment]
Sorry for not knowing this but how do I statically link x265 to libheif? I used vcpkg but have also used cmake before. I am no expert by any means, though.

**Comment by 0xC0000054** [comment]
> Sorry for not knowing this but how do I statically link x265 to libheif?

Statically linking libheif to its dependencies requires manually building libheif, libde265 and x265.
It has been a while since I have tried it, but if I remember correctly it also requires editing the libheif cmake files to force libheif to use the generated libraries. CMake cannot find them automatically when static linking.

It is much easier to use the DLL files the VCPkg generates.

The Access Violation bug mentioned in previous posts was fixed in libheif 1.10.0.
